src/javascript/app_2/pages/portfolio/portfolio.jsx

```diff
@@ -0,0 +1,315 @@
+import PortfolioCard        from './portfolio_card.jsx';
+import DataTable            from '../../components/elements/data_table.jsx';
+import Tooltip              from '../../components/elements/tooltip.jsx';
+import DAO                  from '../../data/dao';
+import { getAppId }         from '../../../config';
+import ClientBase           from '../../../_common/base/client_base';
+import { formatMoney }      from '../../../_common/base/currency_base';
+import { localize }         from '../../../_common/localize';
+import { getPropertyValue } from '../../../_common/utility';
+import Loading              from '../../../../templates/_common/components/loading.jsx';
+
+const formatPortfolioData = (portfolio_arr) => {
+    const formatted_portfolio = portfolio_arr.map((portfolio_item) => {
+        const moment_obj      = moment.utc(portfolio_item.expiry_time * 1000);
+        const remaining_time  = `${moment_obj.fromNow(true)} - ${moment_obj.format('h:mm:ss')}`;
+        const purchase        = parseFloat(portfolio_item.buy_price);
+        const payout          = parseFloat(portfolio_item.payout);
+        return {
+            reference: {
+                transaction_id: portfolio_item.transaction_id,
+                app_id        : portfolio_item.app_id,
+            },
+            type      : portfolio_item.contract_type,
+            details   : localize(portfolio_item.longcode.replace(/\n/g, '<br />')),
+            purchase  : formatMoney(false, purchase, true),
+            payout    : formatMoney(false, payout, true),
+            remaining_time,
+            id        : portfolio_item.contract_id,
+            indicative: {
+                amount: '',
+                style : '',
+            },
+        };
+    });
+    return formatted_portfolio;
+};
+
+// TODO: move to common
+const contract_type_display = {
+    ASIANU      : 'Asian Up',
+    ASIAND      : 'Asian Down',
+    CALL        : 'Higher',
+    CALLE       : 'Higher or equal',
+    PUT         : 'Lower',
+    DIGITMATCH  : 'Digit Matches',
+    DIGITDIFF   : 'Digit Differs',
+    DIGITODD    : 'Digit Odd',
+    DIGITEVEN   : 'Digit Even',
+    DIGITOVER   : 'Digit Over',
+    DIGITUNDER  : 'Digit Under',
+    EXPIRYMISS  : 'Ends Outside',
+    EXPIRYRANGE : 'Ends Between',
+    EXPIRYRANGEE: 'Ends Between',
+    LBFLOATCALL : 'Close-Low',
+    LBFLOATPUT  : 'High-Close',
+    LBHIGHLOW   : 'High-Low',
+    RANGE       : 'Stays Between',
+    UPORDOWN    : 'Goes Outside',
+    ONETOUCH    : 'Touches',
+    NOTOUCH     : 'Does Not Touch',
+};
+
+// TODO: move to common
+const buildOauthApps = (response) => {
+    if (!response || !response.oauth_apps) return {};
+    const obj_oauth_apps = { 2: 'Binary.com Autoexpiry' };
+    response.oauth_apps.forEach((app) => {
+        obj_oauth_apps[app.app_id] = app.name;
+    });
+    return obj_oauth_apps;
+};
+
+const app_id = getAppId();
+
+class Portfolio extends React.PureComponent  {
+    state = {
+        columns: [
```

Comment:
You can explicitly define `columns` here. I don't see any reason to put `columns` data in `state`. 

The purpose of having data in `state` is for Component class to render / re-render. 

Solution:
1) Define `columns` array after or before 
const  columns  = [...];
2) call columns rather than `this.state.columns`.

Reply:
The columns are being rendered / re-rendered based on state (from the streaming calls ```subscribeTransaction``` and ```subscribeProposalOpenContract```), so I think they should be in state? :)

Reply:
Because you never `setState` which means, this `columns` prop is not directly related to this Component. So I thought we need to pull out the columns `variable` from `state =`. 

I gave you a wrong suggestion and here is a new suggestion.
Solution

0) Define `columns` array to as a constant.
1) In DataTable, we can put `componentWillReceiveProps` or `shouldComponentUpdate` to listen to props change. 

Major reason:
1) Having multiple `state` will slow down the PageSpeed dramatically. Why do we have to render `Portfolio` rather than `DataTable` component?

- Let's have some talks in our Lunch time :)

Reply:
Oh no never mind first comment is correct! We can move columns outside of state and keep it as a normal property of Portfolio since columns not rendered / updated over time. The property that handles rendering of the table is data_source